### PR TITLE
Use std::optional for host types

### PR DESCRIPTION
**Pull request by robertmaynard**
## Description
cuda::std::optional shouldn't be used for host types such as `std::vector` as it requires the constructors of the `T` types to be host+device.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by bdice — APPROVED** [review]

File: cpp/src/io/parquet/reader_impl_chunking.cu [hunk 1]
```diff
@@ -386,7 +386,7 @@ int64_t find_next_split(int64_t cur_pos,
     // if decimal but not outputting as float or decimal, then convert to no logical type
     if (column_type_id != type_id::FLOAT64 and
         not cudf::is_fixed_point(data_type{column_type_id})) {
-      return std::make_tuple(clock_rate, cuda::std::nullopt);
+      return std::make_tuple(clock_rate, std::nullopt);
```
Comment (mythrocks) [inline cpp/src/io/parquet/reader_impl_chunking.cu:389]: While we're here, ought we to consider this too?
```suggestion
      return s{clock_rate, std::nullopt};
```

**Review by mythrocks — APPROVED** [review]
LGTM!

File: cpp/src/io/parquet/reader_impl_chunking.cu (hunk 1 above)
Comment (robertmaynard) [reply cpp/src/io/parquet/reader_impl_chunking.cu]: Done!

**Review by ttnghia — APPROVED** [review]

**Review by bdice — APPROVED** [review]
Re-approving additional changes.

**Comment by bdice** [comment]
There are some failures in CI like this:

```
In file included from $SRC_DIR/cpp/build/_deps/gtest-src/googletest/include/gtest/gtest-printers.h:115,
                 from $SRC_DIR/cpp/build/_deps/gtest-src/googletest/include/gtest/gtest-matchers.h:48,
                 from $SRC_DIR/cpp/build/_deps/gtest-src/googletest/include/gtest/internal/gtest-death-test-internal.h:47,
                 from $SRC_DIR/cpp/build/_deps/gtest-src/googletest/include/gtest/gtest-death-test.h:43,
                 from $SRC_DIR/cpp/build/_deps/gtest-src/googletest/include/gtest/gtest.h:65,
                 from $SRC_DIR/cpp/build/_deps/gtest-src/googlemock/include/gmock/internal/gmock-internal-utils.h:50,
                 from $SRC_DIR/cpp/build/_deps/gtest-src/googlemock/include/gmock/gmock-actions.h:145,
                 from $SRC_DIR/cpp/build/_deps/gtest-src/googlemock/include/gmock/gmock.h:56,
                 from $SRC_DIR/cpp/include/cudf_test/cudf_gtest.hpp:19,
                 from $SRC_DIR/cpp/include/cudf_test/base_fixture.hpp:19,
                 from $SRC_DIR/cpp/tests/io/parquet_common.hpp:19,
                 from $SRC_DIR/cpp/tests/io/parquet_writer_test.cpp:17:
$SRC_DIR/cpp/tests/io/parquet_writer_test.cpp:1033:84: error: invalid initialization of reference of type 'const cuda::std::__4::optional<cudf::io::parquet::detail::ConvertedType>&' from expression of type 'const std::optional<cudf::io::parquet::detail::ConvertedType>'
 1033 |       EXPECT_TRUE(compare_binary(ci.max_values[0], stats.max_value.value(), ptype, ctype) >= 0);
      |                                                                                    ^~~~~
In file included from $SRC_DIR/cpp/tests/io/parquet_writer_test.cpp:17:
$SRC_DIR/cpp/tests/io/parquet_common.hpp:175:93: note: in passing argument 4 of 'int32_t compare_binary(const std::vector<unsigned char>&, const std::vector<unsigned char>&, cudf::io::parquet::detail::Type, const cuda::std::__4::optional<cudf::io::parquet::detail::ConvertedType>&)'
  175 |                        cuda::std::optional<cudf::io::parquet::detail::ConvertedType> const& ctype);
      |                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
```

**Comment by robertmaynard** [comment]
> There are some failures in CI like this:
> 
> ```
> In file included from $SRC_DIR/cpp/build/_deps/gtest-src/googletest/include/gtest/gtest-printers.h:115,
>                  from $SRC_DIR/cpp/build/_deps/gtest-src/googletest/include/gtest/gtest-matchers.h:48,
>                  from $SRC_DIR/cpp/build/_deps/gtest-src/googletest/include/gtest/internal/gtest-death-test-internal.h:47,
>                  from $SRC_DIR/cpp/build/_deps/gtest-src/googletest/include/gtest/gtest-death-test.h:43,
>                  from $SRC_DIR/cpp/build/_deps/gtest-src/googletest/include/gtest/gtest.h:65,
>                  from $SRC_DIR/cpp/build/_deps/gtest-src/googlemock/include/gmock/internal/gmock-internal-utils.h:50,
>                  from $SRC_DIR/cpp/build/_deps/gtest-src/googlemock/include/gmock/gmock-actions.h:145,
>                  from $SRC_DIR/cpp/build/_deps/gtest-src/googlemock/include/gmock/gmock.h:56,
>                  from $SRC_DIR/cpp/include/cudf_test/cudf_gtest.hpp:19,
>                  from $SRC_DIR/cpp/include/cudf_test/base_fixture.hpp:19,
>                  from $SRC_DIR/cpp/tests/io/parquet_common.hpp:19,
>                  from $SRC_DIR/cpp/tests/io/parquet_writer_test.cpp:17:
> $SRC_DIR/cpp/tests/io/parquet_writer_test.cpp:1033:84: error: invalid initialization of reference of type 'const cuda::std::__4::optional<cudf::io::parquet::detail::ConvertedType>&' from expression of type 'const std::optional<cudf::io::parquet::detail::ConvertedType>'
>  1033 |       EXPECT_TRUE(compare_binary(ci.max_values[0], stats.max_value.value(), ptype, ctype) >= 0);
>       |                                                                                    ^~~~~
> In file included from $SRC_DIR/cpp/tests/io/parquet_writer_test.cpp:17:
> $SRC_DIR/cpp/tests/io/parquet_common.hpp:175:93: note: in passing argument 4 of 'int32_t compare_binary(const std::vector<unsigned char>&, const std::vector<unsigned char>&, cudf::io::parquet::detail::Type, const cuda::std::__4::optional<cudf::io::parquet::detail::ConvertedType>&)'
>   175 |                        cuda::std::optional<cudf::io::parquet::detail::ConvertedType> const& ctype);
>       |                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
> ```
> 
> Some additional changes are probably needed.

Thanks. Found the function that needed updating to `std::optional`

**Pull request merged**